### PR TITLE
Link Prediction using graph-aware fine-tuning of LLMs

### PR DESCRIPTION
**Pull request by GentleZhu**
*Issue #, if available:*

*Description of changes:*
This PR supports link prediction using GPEFT with random negatives. It relies on PR #758 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 1]
```diff
@@ -59,9 +61,34 @@ def get_lm_node_feats(g, lm_feat_names, lm_ntypes):
...
 
-class LLMGraphModel(gsmodel.GSgnnNodeModelBase):
-    """ LLMGNN Model Class
+    def require_cache_embed(self):
+        return False
+    
+    def forward(self, input_feats, input_nodes):
+        return input_feats
+
+class CosineRandomNegative(nn.Module):
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:74]: why does it have cosine in the name?

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 2]
```diff
@@ -59,9 +61,34 @@ def get_lm_node_feats(g, lm_feat_names, lm_ntypes):
...
+    def __init__(self):
+        super(CosineRandomNegative, self).__init__()
+        self.margin = 0.5
+
+    def forward(self, pos_score, neg_score):
+        pos_distance = []
+        neg_distance = []
+        for k in pos_score:
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:82]: do you iterate over each element here? why not implement it in a vector way?

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 3]
```diff
@@ -89,7 +116,7 @@ def __init__(self, g, node_lm_configs, h_dim, out_dim, num_layers,
                  use_norm = True,   # use normalization or not, default is True
                  alpha_l2norm = 0,
                  lr = 0.001):
-        super(LLMGraphModel, self).__init__()
+        super(GNNLLM_NC, self).__init__()
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:119]: GPEFT should also be used here.

File: examples/peft_llm_gnn/llm_gnn_model.py [hunk 4]
```diff
@@ -192,10 +221,169 @@ def create_optimizer(self):
...
+                 alpha_l2norm = 0,
+                 lr = 0.001):
+        super(GNNLLM_LP, self).__init__()
+        
+        self.target_etype=target_etype
+        self.alpha_l2norm = alpha_l2norm
+        self.lr = lr
+        self.gnn_encoder = GPEFT(g, h_dim, num_layers, target_ntype, node_lm_configs)
+        # dummy node input encoder to enable LP Infer
+        self.node_input_encoder = DummyNodeInputLayer()
```
Comment (zheng-da) [inline examples/peft_llm_gnn/llm_gnn_model.py:349]: do you have to set up this field?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 1 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:74]: Originally, the normalization is done in the loss function. Now it's in the model computation.

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 2 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:82]: It's a dict of {node_type: scores} as output by the LinkPredictDotDecoder

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 3 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:119]: I intend to do this in a seperate PR that makes the previous node classification uses GPEFT component. Is this okay?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 1 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:74]: I added a assert to ensure this

Comment (zheng-da) [reply examples/peft_llm_gnn/llm_gnn_model.py:74]: i don't know what you mean here. do you mean you normalize the embeddings? so it's equivalent to compute cosine similarity?

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 3 above)
Comment (zheng-da) [reply examples/peft_llm_gnn/llm_gnn_model.py:119]: if you can remember to do so. it'll be better to do it in the same PR. it'll be more manageable.

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 4 above)
Comment (zheng-da) [reply examples/peft_llm_gnn/llm_gnn_model.py:350]: where is the field used? i searched for it. it seems this is the only place where it is used.

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 1 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:74]: yes, cosine similarity is equivalent as dot-product with normalized vectors. And we use LinkPredictDotDecoder here.

File: examples/peft_llm_gnn/llm_gnn_model.py (hunk 4 above)
Comment (GentleZhu) [reply examples/peft_llm_gnn/llm_gnn_model.py:350]: It's here. If we don't have model.node_input_encoder, it will report exception. 
https://github.com/awslabs/graphstorm/blob/b38259de64877455f8dc965582a5dc180eda807d/python/graphstorm/model/gnn.py#L985C35-L985C54

**Review by zheng-da — APPROVED** [review]
